keps/sig-storage/1495-volume-populators/README.md

```diff
@@ -0,0 +1,492 @@
+feature gates for each object type. This approach won't scale to generic populators, which
+will by their nature be too numerous and varied.
+
+This proposal recommends that we relax validation (in core Kubernetes) on the `DataSource`
+field to allow arbitrary object types to be data sources, and rely on new controllers to
+perform the validation on those data sources and provide feedback to users. We will introduce
+a new CRD to register valid datasource kinds, and individual volume populators will
+create a CR for each kind of data source that it understands.
+
+## Motivation
+
+### Goals
+
+- Enable users to create pre-populated volumes in a manner consistent with current practice
+- Enable developers to innovate with new an interesting data sources
+- Avoid changing existing workflows and breaking existing tools with new behavior
+- Ensure that users continue to get accurate error messages and feedback about volume
+  creation progress, whether using existing methods of volume creation or specifying a
+  data source associated with a volume populator.
+- Maintain existing pattern of allowing things to happen out of order. As long as there's
+  a chance things will succeed in the future, just provide a status update and retry
+  again later.
+
+### Non-Goals
+
+- This proposal DOES NOT recommend any specific approach for data populators. The specific
+  design of how a data populator should work will be handled in a separate KEP.
+
+## Proposal
+
+Validation for the `DataSource` field will be relaxed so that any API object (except
+core API objects other than PVCs) may be specified. For objects that are not valid data
+source, rather than relying the the API server to reject them or ignore them (the prior
+behavior was to ignore them by clearing the field) allow them and provide feedback to
+API users with events.
+
+A new controller will be introduced called `data-source-validator` which will watch PVCs
+objects and generate events on objects with data sources that are not considered valid. To
+determine which data sources are valid a new CRD will be introduced valled `VolumePopulator`
+which will allow actual data populator controllers to register the API `groups`/`kinds` that
+they understand with the `data-source-validator` controller. If a PVC has a data source that
+matches the `group`/`kind` of a registered `VolumePopulator` the controller will consider
+it valid and emit no events, relying on the actual populator to give feedback to the user.
+
+Instances of the VolumePopulator CRD will look like:
+
+```
+kind: VolumePopulator
+apiVersion: populator.storage.k8s.io/v1alpha1
+metadata:
+  name: example-populator
+sourceKind:
+  group: example.storage.k8s.io
+  kind: Example
+```
+
+In this model there will be 4 sources of feedback to the end user:
+1. The API server will reject core objects other than PVCs with it current behavior, which
+   is to accept the PVC but to blank out the data source. This is backwards compatible, and
```

Comment:
Why is it OK to wipe and ignore user input?  This is almost NEVER the approach we follow elsewhere

Reply:
This is not my proposal, this is how it currently works, and it's been like that since we added data sources for PVCs.

Reply:
Right, the behavior with the feature gate enabled is to stop dropping the fields, except in the corner case where the data source is a core data type other than a PVC (api group is empty string), because we know that core data types can't be PVC data sources.

Reply:
What I am asking is why it is OK to *ever* ignore user input for an enabled feature?  Am I missing something?  If I specify a data source of a pod and you silently drop that request, I am unhappy.  Wouldn't it always be better to either synchronously fail or retain the config and async fail?

Reply:
We can merge and iterate on this, but I DO want to follow-up.

if we don't allow populators for core types, that's maybe OK but you still shouldn't wipe out the data and pretend the user didn't make a mistake.